### PR TITLE
Add both widgets bookmarks and search 

### PR DESCRIPTION
**Pull request by nyitgroup**
Attached is a new pull request with both widgets as you suggested in #458

### CODE REVIEW

File: app/src/main/java/com/quran/labs/androidquran/SearchWidget.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * Implementation of App Widget functionality.
+ */
+public class SearchWidget extends AppWidgetProvider {
+
+  @Override
+  public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
+    // There may be multiple widgets active, so update all of them
+    for (int appWidgetId : appWidgetIds) {
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/SearchWidget.java]: I am not sure about footprint on memory, but this for loop looks neater than the one used in `BookmarksWidget.onUpdate()`.

File: app/src/main/java/com/quran/labs/androidquran/SearchWidget.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+      intent = new Intent(context, SearchActivity.class);
+      pendingIntent = PendingIntent.getActivity(context, 0, intent, 0);
+      widget.setOnClickPendingIntent(R.id.SearchWidgetBtnSearch, pendingIntent);
+
+      intent = new Intent(context, PagerActivity.class);
+      pendingIntent = PendingIntent.getActivity(context, 0, intent, 0);
+      widget.setOnClickPendingIntent(R.id.SearchWidgetBtnGoToQuran, pendingIntent);
+
+
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/SearchWidget.java]: Unnecessary empty line.

File: app/src/main/java/com/quran/labs/androidquran/SearchWidget.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+      appWidgetManager.updateAppWidget(appWidgetId, widget);
+
+    }
+    super.onUpdate(context, appWidgetManager, appWidgetIds);
+
+  }
+
+  @Override
+  public void onEnabled(Context context) {
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/SearchWidget.java]: f it is not a required method to override, I'd prefer you remove it as there is no implementation

File: app/src/main/java/com/quran/labs/androidquran/SearchWidget.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    // Enter relevant functionality for when the first widget is created
+  }
+
+  @Override
+  public void onDisabled(Context context) {
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/SearchWidget.java]: Ditto.

File: app/src/main/java/com/quran/labs/androidquran/SearchWidget.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  }
+  @Override
+  public void onReceive(Context context, Intent intent) {
+    super.onReceive(context, intent);
+
+    updateWidget(context);
+
+  }
+
+  public static void updateWidget(Context context) {
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/SearchWidget.java]: Why public, and why it is an empty method? Please remove.

File: app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java [hunk 6]
```diff
@@ -143,6 +149,8 @@ public boolean replaceRecentRangeWithPage(int deleteRangeStart, int deleteRangeE
   }
 
   public boolean addRecentPage(int page) {
+
+     BookMarksWidget.updateWidget(context);
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java]: This looks wrong to me. What if there are no widgets installed at all?

File: app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java [hunk 7]
```diff
@@ -233,12 +241,15 @@ public void bulkDelete(List<Long> tagIds, List<Long> bookmarkIds, List<Pair<Long
     } finally {
       mDb.endTransaction();
     }
+    BookMarksWidget.updateWidget(context);
   }
 
+
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java]: Unnecessary change, please revert.

File: app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java [hunk 8]
```diff
@@ -233,12 +241,15 @@ public void bulkDelete(List<Long> tagIds, List<Long> bookmarkIds, List<Pair<Long
...
   public long addBookmarkIfNotExists(Integer sura, Integer ayah, int page) {
     long bookmarkId = getBookmarkId(sura, ayah, page);
     if (bookmarkId < 0) {
       bookmarkId = addBookmark(sura, ayah, page);
+
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java]: Ditto.

File: app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java [hunk 9]
```diff
@@ -248,12 +259,15 @@ public long addBookmark(Integer sura, Integer ayah, int page) {
     values.put(BookmarksTable.SURA, sura);
     values.put(BookmarksTable.AYAH, ayah);
     values.put(BookmarksTable.PAGE, page);
+    BookMarksWidget.updateWidget(context);
     return mDb.insert(BookmarksTable.TABLE_NAME, null, values);
   }
 
+
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java]: Ditto.

File: app/src/main/java/com/quran/labs/androidquran/ui/PagerActivity.java [hunk 10]
```diff
@@ -3,6 +3,7 @@
 import android.annotation.TargetApi;
 import android.app.ProgressDialog;
 import android.app.SearchManager;
+import android.appwidget.AppWidgetManager;
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/ui/PagerActivity.java]: Unused import? It looks like there is no significant change in this file at all. Please revert.

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+import android.widget.RemoteViewsService;
+import com.quran.labs.androidquran.R;
+import com.quran.labs.androidquran.dao.Bookmark;
+import com.quran.labs.androidquran.data.QuranInfo;
+import com.quran.labs.androidquran.database.BookmarksDBAdapter;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Created by adnan on 4/1/2017.
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: Auto-generated comment, please remove.

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+  }
+
+  @Override
+  public void onCreate() {
+
+  }
+
+  @Override
+  public void onDataSetChanged() {
+populateListItem();
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: The line is not properly aligned.

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+
+  }
+
+  @Override
+  public void onDestroy() {
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: If this is not a required method, please remove.

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+      item.page = bm.page;
+
+      listItemList.add(item);
+    }
+
+
+  }
+
+  @Override
+  public void onCreate() {
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: If this is not a required method, please remove.

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+
+    for (int i = 0; i < bookmarksList.size(); i++) {
+      Bookmark bm = bookmarksList.get(i);
+      ItemList item = new ItemList();
+      item.sura = QuranInfo.getSuraNameFromPage(appContext,bm.page);
+      item.page = bm.page;
+
+      listItemList.add(item);
+    }
+
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: Unnecessary empty line.

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+  *
+  */
+  @Override
+  public RemoteViews getViewAt(int position) {
+
+    final RemoteViews remoteView = new RemoteViews(
+        context.getPackageName(), R.layout.list_row);
+    ItemList item = listItemList.get(position);
+
+//    remoteView.setTextViewText(R.id.widget_item, item.sura+"-"+item.page);
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: Please remove this line.

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+    remoteView.setTextViewText(R.id.Suratitle, item.sura+"");
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: Please put space before and after plus sign.

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+    remoteView.setTextViewText(R.id.Surametadata, context.getResources().getText(R.string.quran_page)+" "+item.page);
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: Ditto.

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+    remoteView.setImageViewResource(R.id.WidgetFavIcon, R.drawable.ic_favorite);
+    Bundle extras = new Bundle();
+    extras.putInt("page", item.page);
+
+
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: Unnecessary empty line.

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+    Intent fillInIntent = new Intent();
+    fillInIntent.putExtras(extras);
+    remoteView.setOnClickFillInIntent(R.id.widget_item, fillInIntent);
+
+    // interim.
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: Please remove.

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+
+
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: Unnecessary empty line.

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+  public RemoteViews getLoadingView() {
+    return null;
+  }
+
+  @Override
+  public int getViewTypeCount() {
+    return 1;
+  }
+
+
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java]: Ditto.

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetService.java [hunk 23]
```diff
@@ -0,0 +1,28 @@
+package com.quran.labs.androidquran.widgets;
+
+import android.appwidget.AppWidgetManager;
+import android.content.Intent;
+import android.widget.RemoteViewsService;
+
+/**
+ * Created by adnan on 4/1/2017.
```
Comment (ozbek) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetService.java]: Auto-generated comment, please remove.

File: app/src/main/res/layout/book_marks_widget.xml [hunk 24]
```diff
@@ -0,0 +1,67 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (ozbek) [inline app/src/main/res/layout/book_marks_widget.xml]: "bookmarks" is a single word, please do not separate it.

File: app/src/main/res/layout/book_marks_widget.xml [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+        android:layout_height="match_parent"
+        android:layout_weight="1"
+        android:background="@color/translation_verse_text_color"
+        android:src="@drawable/ic_goto_quran" />
+
+  </LinearLayout>
+
+  <LinearLayout
+      android:layout_width="match_parent"
+      android:layout_height="2px"
```
Comment (ozbek) [inline app/src/main/res/layout/book_marks_widget.xml]: Why px here?

File: app/src/main/res/layout/book_marks_widget.xml [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+      android:background="@color/ayah_bookmark_color" />
+
+  <ListView
+      android:id="@+id/listViewWidget"
+      android:layout_width="match_parent"
+      android:layout_height="match_parent"
+      android:background="@color/transparent_actionbar_color" />
+
+
+  <!-- Empty view is show if list items are empty -->
```
Comment (ozbek) [inline app/src/main/res/layout/book_marks_widget.xml]: show -> shown

File: app/src/main/res/layout/book_marks_widget.xml [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+  <TextView
+      android:id="@+id/empty_view"
+      android:layout_width="match_parent"
+      android:layout_height="wrap_content"
+      android:gravity="center"
+      android:text="Bookmarks"
```
Comment (ozbek) [inline app/src/main/res/layout/book_marks_widget.xml]: Please do not use string literal, use the one provided in strings.xml.

File: app/src/main/res/layout/list_row.xml [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+      android:layout_width="match_parent"
+      android:layout_height="wrap_content"
+      android:orientation="vertical">
+
+    <TextView
+        android:id="@+id/Suratitle"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginBottom="4dp"
+        android:text="Al-Baqara"
```
Comment (ozbek) [inline app/src/main/res/layout/list_row.xml]: I understand this is going to be replaced, but please remove it. For tooling purposes, please use `tools:text`

File: app/src/main/res/layout/list_row.xml [hunk 29]
```diff
@@ -0,0 +1,44 @@
...
+        android:textAppearance="@style/TitleText" />
+
+    <TextView
+        android:id="@+id/Surametadata"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="Makki - 286 Verses"
```
Comment (ozbek) [inline app/src/main/res/layout/list_row.xml]: Ditto.

File: app/src/main/res/layout/list_row.xml [hunk 30]
```diff
@@ -0,0 +1,44 @@
...
+        android:textColor="@color/sura_details_color"
+        android:textSize="14sp" />
+
+  </LinearLayout>
+
+
```
Comment (ozbek) [inline app/src/main/res/layout/list_row.xml]: Unnecessary empty line.

File: app/src/main/res/layout/search.xml [hunk 31]
```diff
@@ -11,16 +11,43 @@
...
+      android:layout_height="wrap_content"
+      android:orientation="horizontal">
+
+    <ImageView
+        android:id="@+id/searchSubmit"
+        android:layout_width="wrap_content"
+        android:layout_height="match_parent"
+        android:layout_weight="1"
+        android:src="@drawable/ic_ab_search"
+        android:text="" />
```
Comment (ozbek) [inline app/src/main/res/layout/search.xml]: Please remove.

File: app/src/main/res/layout/search.xml [hunk 32]
```diff
@@ -11,16 +11,43 @@
...
+
+    <EditText
+        android:id="@+id/searchEditText"
+        android:layout_width="328dp"
+        android:layout_height="wrap_content"
+        android:ems="10"
+        android:hint="@string/search_hint"
+        android:inputType="textPersonName"
+        android:text="" />
```
Comment (ozbek) [inline app/src/main/res/layout/search.xml]: Ditto.

File: app/src/main/res/layout/search_widget.xml [hunk 33]
```diff
@@ -0,0 +1,35 @@
...
+
+    <ImageButton
+        android:id="@+id/SearchWidgetBtnGoToQuran"
+        android:layout_width="0dp"
+        android:layout_height="match_parent"
+        android:layout_weight="1"
+        android:background="@color/translation_verse_text_color"
+        android:src="@drawable/ic_goto_quran" />
+
+
```
Comment (ozbek) [inline app/src/main/res/layout/search_widget.xml]: Unnecessary empty line.

File: app/src/main/res/values-v14/dimens.xml [hunk 34]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+
+  <!--
+Refer to App Widget Documentation for margin information
```
Comment (ozbek) [inline app/src/main/res/values-v14/dimens.xml]: Please align the line.

File: app/src/main/res/values/dimens.xml [hunk 35]
```diff
@@ -36,4 +36,10 @@
   <dimen name="tag_width">50dp</dimen>
   <dimen name="tag_margin">6dp</dimen>
   <dimen name="tag_text_size">10sp</dimen>
+
+  <!--
+Refer to App Widget Documentation for margin information
```
Comment (ozbek) [inline app/src/main/res/values/dimens.xml]: Ditto.

File: app/src/main/res/values/strings.xml [hunk 36]
```diff
@@ -366,4 +366,7 @@
 
   <!-- format strings -->
   <string name="sura_ayah" translatable="false">%1$d:%2$d</string>
+  <string name="bookmarks_appwidget_text">Quran BookMarks</string>
```
Comment (ozbek) [inline app/src/main/res/values/strings.xml]: Bookmarks

File: app/src/main/res/xml/book_marks_widget_info.xml [hunk 37]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<appwidget-provider xmlns:android="http://schemas.android.com/apk/res/android"
+    android:initialKeyguardLayout="@layout/book_marks_widget"
+    android:initialLayout="@layout/book_marks_widget"
+    android:minHeight="110dp"
+    android:minWidth="180dp"
+    android:previewImage="@drawable/bookmarks_appwidget_preview"
+    android:resizeMode="horizontal|vertical"
+    android:updatePeriodMillis="5000"
```
Comment (ozbek) [inline app/src/main/res/xml/book_marks_widget_info.xml]: This is too frequent, please put an appropriate period. Anything below 30 mins is going to get ignored anyway.

File: app/src/main/res/xml/search_widget_info.xml [hunk 38]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<appwidget-provider xmlns:android="http://schemas.android.com/apk/res/android"
+    android:initialKeyguardLayout="@layout/search_widget"
+    android:initialLayout="@layout/search_widget"
+    android:minHeight="40dp"
+    android:minWidth="180dp"
+    android:previewImage="@drawable/search_appwidget_preview"
+    android:resizeMode="horizontal|vertical"
+    android:updatePeriodMillis="86400000"
+    android:widgetCategory="home_screen"></appwidget-provider>
```
Comment (ozbek) [inline app/src/main/res/xml/search_widget_info.xml]: You could simply use a closing brace, `/>`, and remove the closing tag, `</appwidget-provider>`.

File: app/src/main/res/xml/search_widget_info.xml [hunk 39]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<appwidget-provider xmlns:android="http://schemas.android.com/apk/res/android"
+    android:initialKeyguardLayout="@layout/search_widget"
+    android:initialLayout="@layout/search_widget"
+    android:minHeight="40dp"
+    android:minWidth="180dp"
+    android:previewImage="@drawable/search_appwidget_preview"
+    android:resizeMode="horizontal|vertical"
+    android:updatePeriodMillis="86400000"
```
Comment (ozbek) [inline app/src/main/res/xml/search_widget_info.xml]: Why 24 hours? :) I'd personally disable the update at all, because this does not seem like an interactive widget.

**Comment by nyitgroup** [comment]
I appreciate you taking the time to look at all those empty lines, comments, unused implementations, and typos. 
That was very helpful. 

I guess I am a little too quick and the IDE did most of the damage. :)

About the ui I am not sure how to make it more elegant. You know some of us are not artists. So be a little more specific example alignment font padding etc. 

Thank you,
Adnan

**Comment by ozbek** [comment]
> About the ui I am not sure how to make it more elegant. You know some of us are not artists. So be a little more specific example alignment font padding etc.

Same here, non-designer :) If you think it can not be improved, then I have no specific comments. That was just a suggestion/question.
